proposals/hasCrossSiteAncestor.md

```diff
@@ -0,0 +1,91 @@
+
+#### Use Cases
+
+The methods used to get, set, remove and getAll cookies.
+
+### Known Consumers
+All adopters of partitioned cookies
+
+## Specification
+
+### Schema
+
+Adds a new optional boolean value hasCrossSiteAncestor to the partitionKey of the cookies details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.
+
+```
+//Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
+"details" {
+  "partitionKey": {
+    "topLevelSite" : "https://example.com"
+    "hasCrossSiteAncestor" : true
+  }
+}
+```
+
+### New Permissions
+No new permissions are required.
+
+### Manifest File Changes
+No new manifest fields are required.
+
+## Security and Privacy
+Privacy: The hasCrossSiteAncestor boolean value does not reveal private information.
+
+Security: The hasCrossSiteAncestor boolean value does not impact security.
+
+### Exposed Sensitive Data
+No sensitive data is associated with the cross-site ancestor chain bit.
+
+### Abuse Mitigations
+A cookie may only have no cross-site ancestor when the topLevelSite in the partitionKey and the URL with which the cookie is associated with are first party to each other. To prevent the creation of cookies that violate this, the set method will return an error if the URL and the topLevelSite are not first party for cookies that are set with no-cross site ancestor.
+
+### Additional Security Considerations
+None
+
+## Alternatives
+The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.
+
+### Existing Workarounds
+To access cookies with the same topLevelSite but different hasCrossSiteAncestor values (A1->B->A2 context), developers can remove the partitioned attribute from the cookie and use the Storage Access Api. 
+
+### Open Web API
+The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.
+
+## Implementation Notes
+APIs affected by the change and the behavior assoicated with the change:
+
+- `cookies.get()`:
+If there is a partitionKey present and it has a value for topLevelSite but no value for hasCrossSiteAncestor, the get method will try to deduce what the value for hasCrossSiteAncestor is likely to be when getting the cookie. In the event that the value is incorrect, developers can either pass the correct value for hasCrossSiteAncestor or use cookies.getAll() with no hasCrossSiteAncestor value set.
```

Comment:
> Does the following language work for you?
> 
> If a `topLevelSite` and the `url` associated with the cookie are the same, then the `hasCrossSiteAncestor` value will be set to `false` otherwise the value should default to `true`.

Why do you want to have dynamic behavior, instead of always defaulting to false?

From the given examples, it looks like the "cross site ancestor bit" is part of the frame tree.

Reply:
If the `url` and `topLevelSite` aren't the same, the the default value of `false` for the hasCrossSiteAncestor value would be invalid. In the scenario referenced, since no value is provided for `hasCrossSiteAncestor`, the API should not default to an invalid value.

Additionally, ensuring that the value for `hasCrossSiteAncestor` can not be invalid, aligns it with the behavior present in `cookies.set()` which performs a similar check.

Reply:
> The "trying to deduce" part of this logic is underspecified.

+1. I think this can be written a bit more procedurally. It is up to us to determine what "trying to deduce" means here.

> Why do you want to have dynamic behavior, instead of always defaulting to false?

Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.

That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should *never* be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.

On the other hand, if `url` and `topLevelSite` are same-site, then we could have cookies whose key has a `false` `hasCrossSiteAncestor` and cookies whose keys have a `true` value (those set in A>B>A contexts). I think in the `url` and `topLevelSite` being same-site case it is reasonable to default to `false` for `set`, `get`, and `delete`. But users can provide a value manually if they wish to access cookies in A>B>A contexts.

For `getAll` I think it is reasonable to include both values (when it makes sense) in our query for cookies if a value is not provided.

How does that all sound?

Reply:
> > Why do you want to have dynamic behavior, instead of always defaulting to false?
> 
> 
> 
> Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.

Are you advocating for unconditional true, or for automatic deduction?


> That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should *never* be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.

Whether two URLs are same-site is not fixed. When the Public Suffix List (PSL) updates, the same-site-ness of two domains may change. Allowing false enables cookie managers to migrate cookies in this case.

> How does that all sound?

Other than what I commented on, sgtm.

Reply:
> > > Why do you want to have dynamic behavior, instead of always defaulting to false?
> > 
> > 
> > Because, at least for our implementation in Chrome, most cases partitioned cookies actually have this value set to `true` unless they are set by the first-party site without any cross-site ancestors. So really there is only a special case where this is not `true`, hence I think that is probably the more sensible default.
> 
> Are you advocating for unconditional true, or for automatic deduction?
> 
> > That being said, as long as the `url` and `topLevelSite` are cross-site, then `hasCrossSiteAncestor` should _never_ be false. I actually think it might even be wise to throw an error if a user tries to pass `false` in this case.
>>> Having an invalid `hasCrossSiteAncestor` value throw an error makes sense to me. When the value is not provided, it should be automatically calculated. Since defaulting to `true` or `false` without evaluating if those values could be valid, will result in cookies with invalid partition keys. 

> Whether two URLs are same-site is not fixed. When the Public Suffix List (PSL) updates, the same-site-ness of two domains may change. Allowing false enables cookie managers to migrate cookies in this case.
>> To handle this use case, a cookie manager looking to do a migration, would be able to use `getAll()` to obtain all the cookies present (regardless of if their partition keys are valid after a PSL update).  Then the cookie manager could iterate through the cookies and use `set()` to attempt to create cookies with the `hasCrossSiteAncestor` value of `false`. If the `set()` returns an error, the manager will know the `topLevelSite` and the `url` are still cross-site after the PSL update. 
> > How does that all sound?
> 
> Other than what I commented on, sgtm.

Reply:
Resolving because discussion is continuing in more recent [comment].(https://github.com/w3c/webextensions/pull/581#discussion_r1701585197)